### PR TITLE
Feature/additional tags

### PR DESCRIPTION
**Pull request by isen-ng**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

***

Fixes https://github.com/traveloka/terraform-aws-vpc/issues/15

***

Release note for [CHANGELOG](https://github.com/traveloka/terraform-aws-vpc/blob/master/CHANGELOG.md):

```release-note
ENHANCEMENTS:

* feature: Additional, optional variable: "tag" that will be appending to existing tags for resources
```

***

Output from `terraform plan` command from changes you propose.

```
$ terraform apply
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
```

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -203,42 +185,33 @@ resource "aws_nat_gateway" "this" {
...
-  tags = {
-    Name          = "${var.vpc_name}-public-rtb"
-    Tier          = "public"
-    ProductDomain = "${var.product_domain}"
-    Environment   = "${var.environment}"
-    Description   = "Route table for public subnet on ${var.vpc_name} VPC"
-    ManagedBy     = "terraform"
-  }
+  tags = "${merge(
+    map("Name", format("%s-public-public", var.vpc_name)),
```
Comment (rafikurnia) [inline main.tf]: Should be changed to:
```hcl
  map("Name", format("%s-public-rtb", var.vpc_name)),
```

File: main.tf [hunk 2]
```diff
@@ -24,6 +24,12 @@ locals {
   role_name_prefix          = "${format(local.role_name_format, var.vpc_name)}"
   role_name_max_byte_length = "${(local.role_name_max_length - length(local.role_name_prefix)) / "2"}"
   role_name_byte_length     = "${min(local.max_byte_length, local.role_name_max_byte_length)}"
+
+  common_tags = "${merge(
```
Comment (rafikurnia) [inline main.tf]: From [Terraform Interpolation Syntax](https://www.terraform.io/docs/configuration-0-11/interpolation.html#merge-map1-map2-), it clearly defined that `merge` will: 
>  Returns the union of 2 or more maps. The maps are consumed in the order provided, and ***duplicate keys overwrite previous entries.***

We don't want to have those `common_tags` overridden by user using `tags` variable you propose.  
So how about changing the order?

Modify from:
```hcl
  common_tags = "${merge(
    map("ProductDomain", var.product_domain), 
    map("Environment", var.environment), 
    map("ManagedBy", "terraform"), 
    var.tags)}"
```

to:
```hcl
  common_tags = "${merge(var.tags,
    map("ProductDomain", var.product_domain), 
    map("Environment", var.environment), 
    map("ManagedBy", "terraform"))}"
```

File: main.tf [hunk 3]
```diff
@@ -39,14 +45,11 @@ resource "aws_vpc" "this" {
...
 
-  tags = {
-    Name          = "${var.vpc_name}"
-    MultiTier     = "${var.vpc_multi_tier ? "true" : "false"}"
-    ProductDomain = "${var.product_domain}"
-    Environment   = "${var.environment}"
-    Description   = "${var.environment} VPC for ${var.product_domain} product domain"
-    ManagedBy     = "terraform"
-  }
+  tags = "${merge(
```
Comment (rafikurnia) [inline main.tf]: Apply to this one as well.. This one and the rest of similar implementation..

File: variables.tf [hunk 4]
```diff
@@ -48,3 +48,8 @@ variable "flow_logs_log_group_retention_period" {
   description = "Specifies the number of days you want to retain log events in the specified log group."
   default     = "14"
 }
+
+variable "tags" {
```
Comment (rafikurnia) [inline variables.tf]: How about naming this variable as `additional_tags`, what do you think?

File: variables.tf [hunk 5]
```diff
@@ -48,3 +48,8 @@ variable "flow_logs_log_group_retention_period" {
...
+  description = "A map of tags to add to all resources"
+  default     = {}
+}
```
Comment (rafikurnia) [inline variables.tf]: Add a newline. Or you can execute `terraform fmt` to make sure it formatted well

File: main.tf (hunk 1 above)
Comment (isen-ng) [reply main.tf]: Fixed

File: main.tf (hunk 2 above)
Comment (isen-ng) [reply main.tf]: Yes i agree. Changed.

File: main.tf (hunk 3 above)
Comment (isen-ng) [reply main.tf]: yes done

File: variables.tf (hunk 5 above)
Comment (isen-ng) [reply variables.tf]: done

File: variables.tf (hunk 4 above)
Comment (isen-ng) [reply variables.tf]: agreed. changed to `additional_tags`

**Comment by isen-ng** [comment]
@rafikurnia Thank you for your feedback!
made the changes!

File: main.tf [hunk 6]
```diff
@@ -24,6 +24,12 @@ locals {
   role_name_prefix          = "${format(local.role_name_format, var.vpc_name)}"
   role_name_max_byte_length = "${(local.role_name_max_length - length(local.role_name_prefix)) / "2"}"
   role_name_byte_length     = "${min(local.max_byte_length, local.role_name_max_byte_length)}"
+
+  common_tags = "${merge(
+    var.additional_tags
```
Comment (rafikurnia) [inline main.tf]: missing comma here :wink: 
```hcl
    var.additional_tags,
```

Comment (isen-ng) [reply main.tf]: Thank you for spotting it!

Comment (isen-ng) [reply main.tf]: thank you for spotting this!
fixed!

**Pull request merged**